src/App.tsx

```diff
@@ -1,24 +1,134 @@
+      default:
+        return todos;
+    }
+  }, [todos, completionFilter]);
+
+  const addTodo = (newTodo: Todo) => {
+    setTodos(currentTodos => [...currentTodos, newTodo]);
+  };
+
+  const deleteTodo = (todoId: number) => {
+    setTodos(currentTodos => (
+      currentTodos.filter(todo => todo.id !== todoId)
+    ));
+  };
+
+  const deleteCompletedTodos = () => {
+    completedTodos.forEach(completedTodo => {
+      deleteTodoFromServer(completedTodo.id)
+        .then(() => deleteTodo(completedTodo.id))
+        .catch(() => setErrorMessage(TodoError.Delete));
+    });
+  };
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-loading-todos#react-todo-app-load-todos">React Todo App - Load Todos</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <header className="todoapp__header">
+          {/* this buttons is active only if there are some active todos */}
+          {todos.length !== 0 && (
+            <button
+              type="button"
+              className={classNames('todoapp__toggle-all', {
+                active: todos.length === completedTodos.length,
+              })}
+            />
+          )}
+
+          {/* Add a todo on form submit */}
+          <TodoForm
+            addTodo={addTodo}
+            setErrorMessage={setErrorMessage}
+            setTempTodo={setTempTodo}
+          />
+        </header>
+
+        <section className="todoapp__main">
+          <TodoList
+            todos={visibleTodos}
+            tempTodo={tempTodo}
+            deleteTodo={deleteTodo}
+            setErrorMessage={setErrorMessage}
+          />
+        </section>
+
+        {/* Hide the footer if there are no todos */}
+        {todos.length !== 0 && (
+          <footer className="todoapp__footer">
+            <span className="todo-count">
+              {`${activeTodos.length} items left`}
+            </span>
+
+            {/* Active filter should have a 'selected' class */}
+            <TodoFilter
+              completionFilter={completionFilter}
+              setCompletionFilter={setCompletionFilter}
+            />
+
+            {/* don't show this button if there are no completed todos */}
+            <button
+              type="button"
+              className="todoapp__clear-completed"
+              disabled={completedTodos.length === 0}
+              onClick={deleteCompletedTodos}
+            >
+              Clear completed
+            </button>
+          </footer>
+        )}
+      </div>
+
+      {/* Notification is shown in case of any error */}
+      {/* Add the 'hidden' class to hide the message smoothly */}
+      {errorMessage && (
```

Comment:
This check is not necessary, because inside the component, you can pass an ```errorMessage``` and include the ```hidden ```class if ```!errorMessage```,

Reply:
The `componentDidMount` method won't be triggered more than once if we always render this component.